add CIDR allocator for NodeController

mqliang opened the pull request:
This PR:
- use pkg/controller/framework to watch nodes and   reduce lists when allocate CIDR for node
- decouple the cidr allocation logic from monitoring status logic

derekwaynecarr commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
update year

derekwaynecarr commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,175 @@
+
+import (
+	"errors"
+	"k8s.io/kubernetes/pkg/util"
```
nit: traditionally we organize imports in groupings like:

```
"fmt" // standard go packages

"k8s packages"
```

derekwaynecarr commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,175 @@
+)
+
+const (
+	cidrsBufSize         = 16
```
please godoc all constants

derekwaynecarr commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,175 @@
+
+var errCIDRRangeNoCIDRsRemaining = errors.New("CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range")
+
+type CIDRAllocator interface {
```
Godoc please

derekwaynecarr commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,175 @@
+	rand           *rand.Rand
+}
+
+func NewCIDRRangeAllocator(clusterCIDR *net.IPNet, subNetMaskSize int) CIDRAllocator {
```
nit: Godoc

derekwaynecarr commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,175 @@
+	r.used.SetBit(&r.used, int(used), 0)
+}
+
+// fillPorts loops, always searching for the next free cidr and, if found, fills the cidrs buffer with it.
```
fillCDIRs

derekwaynecarr commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -260,17 +277,44 @@ func (nc *NodeController) Run(period time.Duration) {
+		return
+	}
+
+	glog.V(4).Infof("Recyce node %s CIDR %s", node.Name, podCIDR)
```
nit: typo "Recycle"

mqliang commented:
@derekwaynecarr Updated. PTAL

derekwaynecarr commented:
lgtm, thanks

mqliang commented:
@derekwaynecarr rebased.

therc commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+
+var errCIDRRangeNoCIDRsRemaining = errors.New("CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range")
+
+// CIDRAllocator ss an interface implemented by things that know how to allocate/recycle CIDR for nodes.
```
typo: ss -> is

therc commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+	rand           *rand.Rand
+}
+
+// NewCIDRRangeAllocator return a CIDRAllocator to allocate CIDR for node
```
returns

therc commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+
+	ip := make([]byte, 4)
+	j := uint32(nextUnused) << uint32(32-r.subNetMaskSize)
+	ipInt := (binary.BigEndian.Uint32(clusterIP)) | j
```
This would give incorrect (duplicate) results if clusterCIDR is not actually the first address in the range. That kind of mistake shouldn't happen, but someone, somewhere will make a typo. I suppose this should be validated upstream in controllermanager.go and log a warning (while ANDing the IP with the mask) or terminate execution.

mqliang replied:
The `ParseCIDR()` function of golang will AND the IP with the mask when parse the cidr, so the `clusterIP` here is the first address in the range..

derekwaynecarr replied:
is there a test case that demonstrates that?

mqliang replied:
```
// ParseCIDR parses s as a CIDR notation IP address and mask,
// like "192.168.100.1/24" or "2001:DB8::/48", as defined in
// RFC 4632 and RFC 4291.
//
// It returns the IP address and the network implied by the IP
// and mask.  For example, ParseCIDR("192.168.100.1/16") returns
// the IP address 192.168.100.1 and the network 192.168.0.0/16.
func ParseCIDR(s string) (IP, *IPNet, error) 
```

We need a test case to verify the behaviour of golang library?

cjcullen replied:
You could just AND the IP w/ the mask to make your assumption explicit:

```
clusterIP := r.clusterCIDR.IP.Mask(r.clusterCIDR.Mask).To4()
```

mqliang commented:
@gmarek

gmarek commented:
Sure, but I'm currently struggling with some P0 issues - I'm not sure I'll be able to do this this week (and I'm OOO in the next one).

mqliang commented:
@k8s-bot test this please

davidopp commented:
@ArtfulCoder @cjcullen Can you guys take a look at this?

cjcullen commented:
Looking now.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+	nodeNum := 1 << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	nextUnused, err := func() (int, error) {
+		j := r.rand.Intn(nodeNum)
```
Do we actually care if this is random? It makes the logic simpler to just count upwards with your count variable initialized outside of the closure. Right now this is potentially a linear search for each new node (but you'd only notice it when we get close to filling up the space). That isn't too terrible for the # of nodes we're dealing with, but we might as well make it constant-time (and more easy to reason about).

If we actually do get some benefit out of the pseudo-random CIDR assignment, this is fine. You could do it in constant time with some sort of LFSR-style thing, but that feels like over-engineering.

mqliang replied:
I refer to [port_allocator](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/userspace/port_allocator.go) when I write those pseudo-random CIDR assignment stuff.

cjcullen replied:
Can you change it to sequential? Allocation problems are much harder to catch if it is random. Take a look at #20048 for an example of where we missed a problem in review.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+	cidrMaskSize, _ := cidrMask.Size()
+
+	if subNetMaskSize < cidrMaskSize {
+		return &emptyAllocator{}
```
Instead of successfully returning an emptyAllocator, could we make this New() function return an error? We know up front if the allocator has been created with impossible parameters. It feels silly to return something that we know will fail whenever it first gets used.

mqliang replied:
Good point. We could check it at https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L204, where we call `NewNodeController()`.

cjcullen replied:
Did you have a reason that you didn't want to return an error from this function?

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+	clusterMask := r.clusterCIDR.Mask
+	clusterMaskSize, _ := clusterMask.Size()
+
+	nodeNum := 1 << uint32(r.subNetMaskSize-clusterMaskSize)
```
Nit: Can you call this maxCIDRs? On my first read-through, I thought this variable was being used for picking the next CIDR.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+func (r *rangeAllocator) AllocateNext() (cidr *net.IPNet, err error) {
+	select {
+	case cidr = <-r.cidrs:
+	case <-time.After(allocateNextTimeout):
```
If we ever hit the `allocateNextTimeout`, it is a problem with the allocation logic, probably not that we don't have any CIDRs remaining (that would return nil).

I think it would make more sense to have this function's contract be that it returns `cidr, nil` when we have a CIDR to allocate, and `nil, errCIDRRangeNoCIDRsRemaining` when we are actually out of CIDRs.

cjcullen replied:
I just saw that the nil case isn't actually passed on to the channel, and that we treat the resulting hang as the indicator that we have run out. Is there a good reason to use a channel instead of generating on-demand?

mqliang replied:
I refer to port_allocator when I write those channel stuff.

mqliang replied:
> I think it would make more sense to have this function's contract be that it returns cidr, nil when we have a CIDR to allocate, and nil, errCIDRRangeNoCIDRsRemaining when we are actually out of CIDRs.

> I just saw that the nil case isn't actually passed on to the channel

```
func (r *rangeAllocator) AllocateNext() (cidr *net.IPNet, err error) {
```

We declare the variable`cidr` and `err` at function signature, if they don't get assigned in function body, they will be nil, for example: 

```
func test() (a *int, b*int) {
    return
}
``` 

will return `nil, nil`

cjcullen replied:
I'm fine with the channel design if you think it can make the code cleaner.

But, I don't think we should be injecting sleeps unnecessarily (half second between allocations, and a second before returning in the error case).

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+		cidrs:          make(chan *net.IPNet, cidrsBufSize),
+		rand:           rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+	go util.Until(func() { ra.fillCIDRs(util.NeverStop) }, nextFreeCIDRCooldown, util.NeverStop)
```
Why is a cooldown necessary?

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+func (r *rangeAllocator) Release(cidr *net.IPNet) {
+	clusterIP := r.clusterCIDR.IP.To4()
```
Might be nicer to mask this IP here too. Up to you.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,182 @@
+	used := (binary.BigEndian.Uint32(clusterIP) ^ binary.BigEndian.Uint32(cidr.IP.To4())) >> uint32(32-r.subNetMaskSize)
+
+	if used > nodeNum {
+		return
```
We don't expect this case. A log message would be nice.

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -189,6 +191,15 @@ func NewNodeController(
+
+	nodeEventHandlerFuncs := framework.ResourceEventHandlerFuncs{}
+	if nc.allocateNodeCIDRs {
+		nodeEventHandlerFuncs = framework.ResourceEventHandlerFuncs{
```
I'm not very familiar with the ResourceEventHandlers, but I don't think this would handle crashes/restarts well.

E.g., if the nodeController restarts, it is possible that the next new node is issued a CIDR that was previously issued to someone else, right?

mqliang replied:
if the nodeController restarts, ResourceEventHandlers would observe some Add Event. So we need mark the CIDR has been occupied if `node.Spec.PodCIDR != ""`, sorry for the missing.

cjcullen commented:
Separating this logic out from the nodeController is a great step. Thanks!

I dropped some comments. My biggest concern is about the handling of controller-manager restarts.

mqliang committed:
cidr allocator

mqliang commented:
@cjcullen 
Reworked:
1) Add the occupy logic, which will mark the CIDR as used when ControllerManager restarted.
2) Remove the channel and sleep stuff.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,144 @@
+	maxCIDRs = 1 << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	nextUnused, err := func() (int, error) {
+		j := r.rand.Intn(maxCIDRs)
```
I still don't see a good reason for random allocation of CIDRs. I'm willing to be convinced though... :smiley:

mqliang replied:
No good reason. At first, I thought CIDRs will be released randomly, and cause some holes in the CIDR range, search from a random place may help to find a available CIDR quickly. But it just occur to me that it doesn't help a lot for the number of nodes we're dealing with and it makes bug-catch harder, just as you put.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,144 @@
+			return j, nil
+		}
+		// search sequentially
+		for i := j + 1; i < maxCIDRs; i++ {
```
You could condense this into a single loop:
```
for i := 0; i < maxCIDRs; i++ {
	next := (j + i) % maxCIDRs
	if r.used.Bit(next) == 0 {
		return next, nil
	}
}
```

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,144 @@
+				return i, nil
+			}
+		}
+		return -1, fmt.Errorf("not found")
```
Just return `errCIDRRangeNoCIDRsRemaining` here instead of constructing an error you know you're going to throw away.

Then below, you can have:
```
if err != nil {
	return nil, err
}

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -136,8 +138,15 @@ func NewNodeController(
+		}
+		mask := clusterCIDR.Mask
+		// TODO: make pod cidr mask size configurable, by now set it as 24
+		if maskSize, _ := mask.Size(); maskSize > 24 {
```
Why do we require that the mask size be <= 24?

cjcullen replied:
Oh I see. Just preventing use of something larger than a /8?

cjcullen commented on pkg/controller/node/cidr_allocator_test.go:
```diff
@@ -0,0 +1,108 @@
+	for i := 0; i < 256; i++ {
+		a.Release(cidrs[i])
+	}
+
```
Can you add a call to `Occupy()` here (and then allocate 255 CIDRs), just so we have a little coverage on that?

cjcullen replied:
Thanks! (this is handled in the TestRangeAllocator_AllocationOccupied test)

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,144 @@
+type rangeAllocator struct {
+	clusterCIDR    *net.IPNet
+	subNetMaskSize int
+	cidrs          chan *net.IPNet
```
You can remove `cidrs`.

cjcullen commented:
Looking very good. I'm still not thrilled about the random allocation, but I might just be gun-shy from some of the problems we've been tracking down recently.

mqliang commented:
@cjcullen 
Reworked:
1) change random allocation to sequential one
2) add unit test for Occupy()

PTAL

cjcullen commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -66,6 +66,7 @@ type CMServer struct {
 
 	ClusterName       string
 	ClusterCIDR       net.IPNet
+	PodCIDRMaskSize   int
```
It looks like you started to make pod CIDR sizes configurable, but then didn't use it anywhere?

mqliang replied:
I'd like to make pod CIDR sizes configurable in another PR.

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -477,6 +528,7 @@ func (nc *NodeController) monitorNodeStatus() error {
 	return nil
 }
 
+/*
```
Go ahead and delete this code if it isn't going to be used.

cjcullen commented:
Marking as 1.2 since this makes our scaling (especially for large clusters) much cleaner.

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -280,17 +305,49 @@ func (nc *NodeController) Run(period time.Duration) {
+	glog.V(4).Infof("Assigning node %s CIDR %s", node.Name, podCIDR)
+	node.Spec.PodCIDR = podCIDR.String()
+	if _, err := nc.kubeClient.Nodes().Update(node); err != nil {
+		nc.recordNodeStatusChange(node, "CIDRAssignmentFailed")
```
If this happens, do we retry?

Can we also log the error?

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -280,17 +305,49 @@ func (nc *NodeController) Run(period time.Duration) {
+
+	podCIDR, err := nc.cidrAllocator.AllocateNext()
+	if err != nil {
+		nc.recordNodeStatusChange(node, "CIDRNotAvailable")
```
Probably need some sort of retry here too.

mqliang replied:
I have no idea how to implement the retry logic, since `AllocateNext()`has search through the range to find available ones. If `AllocateNext()` failed, it will fail again (unless a node be deleted and its podCIDR be released during retry). If we want add the retry logic here, may be my first version of this PR(using channel and timeout) is much better. Do you have any thoughts?

cjcullen replied:
This is fine how it is. It's definitely no worse than what we have now.

gmarek commented:
Actually with @davidopp we think it's too big change to rush for v1.2 (and it was few weeks back). I don't think anything changed. I'm against fast tracking this change.

cjcullen commented:
Ok. If that decision is already made, feel free to remove the milestone (and reassign if you'd like). I'll open an issue with the problems that do need to be fixed for 1.2.

mqliang commented:
@cjcullen Sorry for the delay. I have rebased this PR. Could you please take a look?

cjcullen commented:
Thanks for rebasing. I'll take a look today.

cc @zmerlynn

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,130 @@
+	ra := &rangeAllocator{
+		clusterCIDR:    clusterCIDR,
+		subNetMaskSize: subNetMaskSize,
+		rand:           rand.New(rand.NewSource(time.Now().UnixNano())),
```
rand appears to be unused now.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,130 @@
+func (r *rangeAllocator) Release(cidr *net.IPNet) {
+	used, err := r.getBitforCIDR(cidr)
+	if err != nil {
+		glog.Errorf("failed to release cidr: %v", err)
```
We should not continue in this function if we are going to set an invalid bit.

cjcullen commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,130 @@
+func (r *rangeAllocator) Occupy(cidr *net.IPNet) {
+	used, err := r.getBitforCIDR(cidr)
+	if err != nil {
+		glog.Errorf("failed to mark cidr as occupied :%v", err)
```
Same here. Change this to return an error if its dependency can return an error.

cjcullen commented on pkg/controller/node/cidr_allocator_test.go:
```diff
@@ -0,0 +1,145 @@
+	if !reflect.DeepEqual(cidrs, rcidrs) {
+		t.Fatalf("expected re-allocated cidrs are the same collection")
+	}
+}
```
Can you add a test that explicitly tests the `getBitForCIDR` method?

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -151,8 +155,15 @@ func NewNodeController(
+		}
+		mask := clusterCIDR.Mask
+		// TODO: make pod cidr mask size configurable, by now set it as 24
+		if maskSize, _ := mask.Size(); maskSize > 24 {
```
Don't need to allocate a new variable just to check it.

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -151,8 +155,15 @@ func NewNodeController(
+			glog.Fatal("NodeController: Must specify clusterCIDR if allocateNodeCIDRs == true.")
+		}
+		mask := clusterCIDR.Mask
+		// TODO: make pod cidr mask size configurable, by now set it as 24
```
Change this TODO to:
```
		// TODO: Make pod CIDR mask size configurable.
		// For now, we assume podCIDR mask size is 24, so make sure the
		// clusterCIDR mask size is larger than 24.

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -220,6 +241,12 @@ func NewNodeController(
 	)
+
+	if allocateNodeCIDRs {
+		// TODO: make pod CIDR mask size configurable, by now set as 24
```
```
		// TODO: make pod CIDR mask size configurable, for now set it to 24.
```

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -295,17 +322,61 @@ func (nc *NodeController) Run(period time.Duration) {
+			glog.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
+			return
+		}
+		nc.cidrAllocator.Occupy(podCIDR)
```
Don't we want to return after this?

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -295,17 +322,61 @@ func (nc *NodeController) Run(period time.Duration) {
+		node.Spec.PodCIDR = podCIDR.String()
+		if _, err := nc.kubeClient.Core().Nodes().Update(node); err == nil {
+			break
+		}
```
Would be nice to log the error from the failed update.

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -295,17 +322,61 @@ func (nc *NodeController) Run(period time.Duration) {
+		}
+		node, err = nc.kubeClient.Core().Nodes().Get(node.Name)
+		if err != nil {
+			glog.Errorf("Failed while getting a Node to retry updating Node.Spec.PodCIDR. Probably Node %s was deleted.", node.Name)
```
Log the error here. We shouldn't assume that the error is due to the node being deleted.

cjcullen commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -295,17 +322,61 @@ func (nc *NodeController) Run(period time.Duration) {
+	}
+}
+
+// recycleCIDR recycle the CIDR of a removed node
```
```
// recycleCIDR recycles ...
```

wojtek-t commented:
@gmarek - can you please also jump into it?

mqliang commented:
Comments addressed. PTAL

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,128 @@
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	clusterIP := r.clusterCIDR.IP.To4()
```
Store all those things in the rangeAllocator instead of computing it during each allocation.

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,128 @@
+	clusterMaskSize, _ := clusterMask.Size()
+
+	var maxCIDRs int
+	maxCIDRs = 1 << uint32(r.subNetMaskSize-clusterMaskSize)
```
Same as above - it doesn't change between calls AFAICT.

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,128 @@
+	nextUnused := -1
+	for i := 0; i < maxCIDRs; i++ {
+		if r.used.Bit(i) == 0 {
+			nextUnused = i
```
break?

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,128 @@
+
+	r.used.SetBit(&r.used, nextUnused, 1)
+
+	ip := make([]byte, 4)
```
I'm not a fan of overoptimizing it with bitwise operator, especially taking into account number of other inefficiencies (e.g. scanning whole 'used' map every time). Please rewrite it into easily readable format e.g. by setting all bytes separately.

mqliang replied:
@gmarek Could you please be more specific here? I am confused about "setting all bytes separately".

gmarek replied:
All bytes of IPv4 address. Just treat it as an array, not BigInt - it's slower and longer, but more readable for most people.

mqliang replied:
It may make the code really complicated if we treat it as an array: I'd like send a follow-up PR to make the podCIDR mask size configureable, if the clusterCIDR mask size and podCIDR mask size  is not multiples of 8, the code here will becomes really complicated and hard to read.

gmarek replied:
They won't be multiples of 8.

mqliang replied:
for example, if I set:
```
--cluster-cidr=10.65.18.0/7
--pod-cidr-size=17
```

treating the ip as []byte will make the code really complicated and hard to read.

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,128 @@
+
+	maxCIDRs := uint32(1) << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	used := (binary.BigEndian.Uint32(clusterIP) ^ binary.BigEndian.Uint32(cidr.IP.To4())) >> uint32(32-r.subNetMaskSize)
```
For readability please treat all bytes separately. People are not use to thinking about IPs as an integer.

gmarek commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -220,6 +243,12 @@ func NewNodeController(
+
+	if allocateNodeCIDRs {
+		// TODO(mqliang): make pod CIDR mask size configurable, for now set it to 24.
+		nc.cidrAllocator = NewCIDRRangeAllocator(clusterCIDR, 24)
```
You need to go through existing Nodes and populate 'usedCIDR' map. Otherwise this won't work after controllerManager restart.

gmarek replied:
It's OK, sorry.

wojtek-t commented:
@mqliang - can you please prioritize this PR? We really really need it. Thanks!

mqliang commented:
@wojtek-t ok, will update tonight.

mqliang commented:
addressed some comments from @gmarek , but still treat IPv4 address as uint32, since treating it as []byte will make the bitwise operation really complicated.

gmarek commented:
I won't block on IPv4 - this is only a preference issue. LGTM

gmarek commented:
One more thing - we should make sure that we don't assign adresses from serviceCIDR.

gmarek commented:
@mqliang - can you add a fix for #24547 to this PR, or write a short follow up?

mqliang commented:
@gmarek IIUC, controller-manager should also have a "service-cluster-ip-range" flag, and when cidr_allocator allocate podcidrs, it should filter out cidrs in service-cluster-ip-range?

wojtek-t commented:
@mqliang - correct

gmarek commented:
@mqliang - as @wojtek-t already wrote - that's correct. We don't have any mechanism that would prevent service range from having nonempty intersection with cluster range, and we can't have services and nodes with the same IP.

mqliang commented:
@wojtek-t @gmarek The last commit try to filter out serviceCIDR overlapping podCIDRs, If the implementation looks good, I will update the auto generation files. BTW, I am not familiar with the cluster installation scripts, I may missed something.

gmarek commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -254,6 +257,9 @@ func NewNodeController(
 
 // Run starts an asynchronous loop that monitors the status of cluster nodes.
 func (nc *NodeController) Run(period time.Duration) {
+
```
remove empty line.

gmarek replied:
ping.

gmarek commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -324,6 +330,19 @@ func (nc *NodeController) Run(period time.Duration) {
+		return
+	}
+
+	for i := 0; i< nc.cidrAllocator.MaxCIDRs(); i++ {
```
This will be extremely inefficient. CIDR range is, in practice an interval of integers, and you're using a bitmask to keep the occupied CIDRs, so you can mark whole service range in a single call.

zmerlynn replied:
In this case, the CIDR allocator is a range of node CIDRs within a cluster CIDR, though, isn't it? They're somewhat well bounded right now, e.g. 32k node CIDRs in a /9 cluster CIDR. Having started to write this PR, this is a similar approach as I took because the logic for going the bitmask exactly right, while fun, takes some crunching (and would take some thorough unit tests), whereas this is plain, slow and minorly hurts startup time only (but 32k is tiny).

gmarek replied:
There's a lot of bit magic in this PR, so I don't think a bit more will hurt:)

zmerlynn replied:
Fair enough. :)

gmarek commented:
@mqliang - one major comment, but looks good otherwise.

mqliang commented:
@wojtek-t @gmarek @zmerlynn Updated. PTAL. Only the last commits need reviewing. If there are some cluster installation scripts need update due to the addition of "--service-cidr" flag, please point out, I am not familiar with those scripts.

zmerlynn commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,153 @@
+	return r.maxCIDRs
+}
+
+func (r *rangeAllocator) Occupy(cidr *net.IPNet) (err error) {
```
Can you add a few more boundary tests for Occupy? In particular, things like the cases where `cidr` exceeds the entire range, or where `cidr` is smaller than a node CIDR. (They can return errors, that's fine.)

gmarek replied:
+1 Please add tests cases that @zmerlynn suggested. Also please add comments to this part of code, as it's by far nontrivial.

zmerlynn commented:
I think you handled the Trusty change (though @andyzheng0831 might want to check). The other leg of this is to modify Salt. You'll need to plumb it in here:
https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest

That CIDR range is available as `pillar['service_cluster_ip_range']` in Salt, you can see an example here: https://github.com/kubernetes/kubernetes/blob/7a09fbcf97243e15ff2d82c4b812d9eee22389df/cluster/saltbase/salt/kube-apiserver/kube-apiserver.manifest

zmerlynn commented:
(There's also several other providers that may want to follow suit, but that covers the GCE ones.)

andyzheng0831 commented:
The change to cluster/gce/trusty/configure-helper.sh lgtm, if you add the new var in kube-controller-manager.manifest in the same way as var service_cluster_ip_range

gmarek commented:
I didn't have time to look at this today. I'll try to do it tomorrow.

gmarek commented:
FAILED   ./hack/../hack/verify-generated-docs.sh. e2e failure is a flake.

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,153 @@
+			IP:   cidr.IP.To4().Mask(mask),
+			Mask: mask,
+		})
+	}
```
else return nil?

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,153 @@
+			Mask: mask,
+		})
+	}
+	if err != nil {
```
move inside the if

mqliang commented:
@gmarek @andyzheng0831 @zmerlynn Updated. Only the last commit need reviewing. If it look good, I will run the scripts in package /hack to update auto generation files.

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+func (r *rangeAllocator) getBitforCIDR(cidr *net.IPNet) (int, error) {
+	used := (binary.BigEndian.Uint32(r.clusterIP) ^ binary.BigEndian.Uint32(cidr.IP.To4())) >> uint32(32-r.subNetMaskSize)
```
I don't understand this expression. Wouldn't it always be 0 after the shift?

wojtek-t replied:
Well - we actually think it's correct after some thinking.

gmarek replied:
s/used/ciderIndex/

zmerlynn commented on cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest:
```diff
@@ -10,6 +10,9 @@
 {% if pillar['cluster_cidr'] is defined and pillar['cluster_cidr'] != "" -%}
   {% set cluster_cidr = "--cluster-cidr=" + pillar['cluster_cidr'] -%}
 {% endif -%}
+{% if pillar['service_cidr'] is defined and pillar['service_cidr'] != "" -%}
```
It's actually `pillar['service_cluster_ip_range']` (I mentioned in https://github.com/kubernetes/kubernetes/pull/19242#issuecomment-218309287). I think that's probably why Jenkins is failing on you as well.

zmerlynn commented on cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest:
```diff
@@ -10,6 +10,9 @@
   {% set cluster_cidr = "--cluster-cidr=" + pillar['cluster_cidr'] -%}
 {% endif -%}
+{% if pillar['service_cidr'] is defined and pillar['service_cidr'] != "" -%}
+  {% set service_cidr = "--cluster-cidr=" + pillar['service_cidr'] -%}
```
And you need to follow the pattern with `cluster_cidr` a little more closely. I don't like this Salt file because the variables are set to nil at the top (rather than right before changing, like a ternary), but you need to add `service_cidr` up near where `cluster_cidr` is set to empty at line 2, and then here on line 14 change `service_cidr` to the new flag. (`--service-cidr=<pillar>`)

zmerlynn replied:
And then way down on line 61, you need to actually use `service_cidr`.

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,154 @@
+	return nil
+}
+
+func (r *rangeAllocator) getBitforCIDR(cidr *net.IPNet) (int, error) {
```
s/Bit/Index/

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,154 @@
+	maskSize, _ := cidrMask.Size()
+
+	if r.clusterCIDR.Contains(cidr.IP.Mask(r.clusterCIDR.Mask)) && r.clusterMaskSize < maskSize {
+		mask := net.CIDRMask(r.subNetMaskSize, 32)
```
s/mask/subNetMask/

gmarek commented:
The Occupy function looks good to me, but I'd like @cjcullen or @zmerlynn to take a look as well. Major parts look good, but maybe @cjcullen want to have another pass.

gmarek commented:
Ping @cjcullen @zmerlynn - this is a blocker for 1.3

gmarek commented:
@mqliang please fix failing tests.

gmarek commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -141,6 +141,7 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.EnableProfiling, "profiling", true, "Enable profiling via web interface host:port/debug/pprof/")
 	fs.StringVar(&s.ClusterName, "cluster-name", s.ClusterName, "The instance prefix for the cluster")
 	fs.StringVar(&s.ClusterCIDR, "cluster-cidr", s.ClusterCIDR, "CIDR Range for Pods in cluster.")
+	fs.StringVar(&s.ServiceCIDR, "service-cidr", s.ServiceCIDR, "CIDR Range for Services cluster.")
```
Reuse "service-cluster-ip-range" flag.

gmarek commented on hack/verify-flags/known-flags.txt:
```diff
@@ -371,6 +371,7 @@ service-account-key-file
 service-account-lookup
 service-account-private-key-file
 service-address
+service-cidr
```
Remove

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,154 @@
+	return nil
+}
+
+func (r *rangeAllocator) getIndexforCIDR(cidr *net.IPNet) (int, error) {
```
s/getIndexforCIDR/getIndexForCIDR/g

gmarek commented on pkg/controller/node/cidr_allocator_test.go:
```diff
@@ -0,0 +1,325 @@
+	}{
+		{
+			clusterCIDRStr:    "127.0.0.0/8",
+			subNetMaskSize:    16,
```
Try something that's not multiples of 8.

gmarek commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -151,8 +157,17 @@ func NewNodeController(
+		// TODO(mqliang): Make pod CIDR mask size configurable.
+		// For now, we assume podCIDR mask size is 24, so make sure the
+		// clusterCIDR mask size is larger than 24.
+		if maskSize, _ := mask.Size(); maskSize > 24 {
```
Where do we need this assumption?

mqliang replied:
It's because the previous hard code the podCIDR mask size as 24, see https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/nodecontroller.go#L303-L313

I'd like make the podCIDR mask size configurable.

gmarek commented:
Added a few more comments, but generally looks good to me. I'd like @zmerlynn or @cjcullen to take a look as well.

mqliang committed:
Don't allow node controller to allocate into service CIDR range

gmarek commented on cluster/gce/trusty/configure-helper.sh:
```diff
@@ -488,6 +488,9 @@ start_kube_controller_manager() {
     params="${params} --cluster-cidr=${CLUSTER_IP_RANGE}"
   fi
+  if [ -n "${SERVICE_IP_RANGE:-}" ]; then
+    params="${params} --service-cidr=${SERVICE_IP_RANGE}"
```
Reuse "service-cluster-ip-range" flag.

gmarek commented on cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest:
```diff
@@ -10,6 +11,9 @@
 {% if pillar['cluster_cidr'] is defined and pillar['cluster_cidr'] != "" -%}
   {% set cluster_cidr = "--cluster-cidr=" + pillar['cluster_cidr'] -%}
 {% endif -%}
+{% if pillar['service_cluster_ip_range'] is defined and pillar['service_cluster_ip_range'] != "" -%}
```
You need to create those pillars in configure-vm.sh

zmerlynn replied:
@gmarek: That pillar is defined. https://github.com/kubernetes/kubernetes/blob/aada051b20c9e47901ca27535cb8a5484bd37870/cluster/gce/configure-vm.sh#L428

gmarek replied:
Ah, yes - another advantage of reusing existing flags;)

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,154 @@
+	cidrMask := cidr.Mask
+	maskSize, _ := cidrMask.Size()
+
+	if r.clusterCIDR.Contains(cidr.IP.Mask(r.clusterCIDR.Mask)) && r.clusterMaskSize < maskSize {
```
if !r.clusterCIDR.Contains(cidr.IP.Mask(r.clusterCIDR.Mask)) || r.clusterMaskSize >= maskSize { return nil } else {...

mqliang replied:
Oh, maybe it should be:
```
	if !r.clusterCIDR.Contains(cidr.IP.Mask(r.clusterCIDR.Mask)) && !cidr.Contains(r.clusterCIDR.IP.Mask(cidr.Mask)){
		return fmt.Errorf("cidr %v is out the range of cluster cidr %v", cidr, r.clusterCIDR)
	}

	if r.clusterMaskSize < maskSize {
		...
	}
```

mqliang replied:
`!r.clusterCIDR.Contains(cidr.IP.Mask(r.clusterCIDR.Mask)) && !cidr.Contains(r.clusterCIDR.IP.Mask(cidr.Mask)) ` indicating clusterCIDR and NodeCIDR/ServiceCIDR not overlapping

`r.clusterMaskSize >= maskSize` indicating that we should occupy the whole range.

gmarek commented on pkg/controller/node/cidr_allocator.go:
```diff
@@ -0,0 +1,154 @@
+			IP:   net.IP(ip).To4().Mask(subNetMask),
+			Mask: subNetMask,
+		})
+
```
remove empty line.

gmarek commented:
Verifying ./hack/../hack/verify-flags-underscore.py
Found flags in golang files not in the list of known flags. Please add these to hack/verify-flags/known-flags.txt
pod-cidr-mask-size

gmarek commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -141,6 +142,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.ClusterName, "cluster-name", s.ClusterName, "The instance prefix for the cluster")
 	fs.StringVar(&s.ClusterCIDR, "cluster-cidr", s.ClusterCIDR, "CIDR Range for Pods in cluster.")
+	fs.StringVar(&s.ServiceCIDR, "service-cluster-ip-range", s.ServiceCIDR, "CIDR Range for Services in cluster.")
+	fs.Int32Var(&s.PodCIDRMaskSize, "pod-cidr-mask-size", s.PodCIDRMaskSize, "Mask size for pod cidr in cluster.")
```
s/PodCIDR/NodeCIDR/g (and all related changes)

zmerlynn commented on cluster/gce/trusty/configure-helper.sh:
```diff
@@ -488,6 +488,9 @@ start_kube_controller_manager() {
     params="${params} --cluster-cidr=${CLUSTER_IP_RANGE}"
   fi
+  if [ -n "${SERVICE_IP_RANGE:-}" ]; then
+    params="${params} --service-cluster-ip-rang=${SERVICE_CLUSTER_IP_RANGE}"
```
`--service-cluster-ip-range`

zmerlynn commented:
@k8s-bot test this, issue #IGNORE. (I want to see the result again to verify, the PR builder seems to be noisy right now.)

mqliang commented:
@zmerlynn @gmarek Great thanks for your patient reviewing. I will squash the commits if looks good.

zmerlynn commented:
@mqliang: It looks like you're still failing unit tests?

```
Generated docs are out of date. Please run hack/update-generated-docs.sh
FAILED   ./hack/../hack/verify-generated-docs.sh	60s
```

gmarek commented:
@mqliang 
```
Verifying ./hack/../hack/verify-flags-underscore.py
Found illegal 'flag' usage. If these are false positives you should run `hack/verify-flags-underscore.py -e > hack/verify-flags/exceptions.txt` to update the list.
cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest:{% set params = "--master=127.0.0.1:8080" + " " + cluster_name + " " + cluster_cidr + " " + allocate_node_cidrs + " " + service_cluster_ip_range + " " + terminated_pod_gc + " " + cloud_provider  + " " + cloud_config + " " + service_account_key + " " + log_level + " " + root_ca_file -%}
FAILED   ./hack/../hack/verify-flags-underscore.py	28s
```

zmerlynn commented:
@mqliang: I realize our timezones are probably mismatched, but we need to get this in ASAP if we can. This LGTM, if you can get past tests. I think you're closest in TZ to @gmarek, so maybe work with him on final approval. @cjcullen, can you give a proxy LGTM as well?

cjcullen commented:
It looks good to me. I'll reassign to @gmarek for final LGTM and application of the label.

gmarek commented:
LTGM once you get the verify-flags-underscore pass. 

Great work by the way - it's a super important change that we were waiting way to long to get in. Thanks!

gmarek commented:
@mqliang 
```
+++ [0519 09:10:23] Starting controller-manager
I0519 09:10:23.137212    6964 plugins.go:71] No cloud provider specified.
I0519 09:10:23.137345    6964 nodecontroller.go:151] Sending events to api server.
I0519 09:10:23.137344    6964 replication_controller.go:240] Starting RC Manager
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x7f1d4b]

goroutine 1 [running]:
panic(0x22f4ea0, 0xc82000e0b0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/controller/node.(*NodeController).filterOutServiceRange(0xc8202877a0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:335 +0x2b
k8s.io/kubernetes/pkg/controller/node.(*NodeController).Run(0xc8202877a0, 0x2540be400)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:262 +0x25
k8s.io/kubernetes/cmd/kube-controller-manager/app.StartControllers(0xc82017c000, 0xc820194d40, 0xc8200904e0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-controller-manager/app/controllermanager.go:235 +0xbf1
k8s.io/kubernetes/cmd/kube-controller-manager/app.Run.func2(0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-controller-manager/app/controllermanager.go:156 +0x4b
k8s.io/kubernetes/cmd/kube-controller-manager/app.Run(0xc82017c000, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-controller-manager/app/controllermanager.go:162 +0x3e5
main.main()
```

Also please retitle the PR, so that the name can be used as a release node.

mqliang commented:
@gmarek  seems no "service-cluster-ip-range" flag passed in, is there anything about the script in /cluster going wrong? I am not familiar with that. Mind to help take a look?

gmarek commented:
@mqliang - I'd love to, but I had to run now. I'll be in the office tomorrow morning (~8am CEST time).

gmarek commented:
It's a `hack/test-cmd.sh` command - you can trace the issue from there.

mqliang commented:
@gmarek Could you please kindly help take a look? I am not familiar with the script in /cluster and test files.

gmarek commented:
Sure. I'll take a look.

gmarek commented:
You can try adding a default for ServiceCIDR in controllerManager options.

mqliang commented:
@gmarek adding a default for ServiceCIDR could help, but the fundamental problems is it seems we didn't pass in the "service-cluster-ip-range flag" in test: ClusterCIDR has no default value, but it didn't bring down test, so we must pass the flag somewhere.

gmarek commented:
Yes, you're right - I guess you're missing clusterCIDR flag in the test-cmd now.

gmarek commented:
Ugh, sorry - I completely forgot that we have a flag that disables CIDR assignment (`allocate-node-cidr=false`). This means that you should gate all new stuff with this flag as well. Then you actually won't need cluster nor service CIDRs.

mqliang commented:
@gmarek So, what should we do now?

gmarek commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -224,11 +248,19 @@ func NewNodeController(
 // Run starts an asynchronous loop that monitors the status of cluster nodes.
 func (nc *NodeController) Run(period time.Duration) {
+
+	nc.filterOutServiceRange()
```
Wrap this in `if allocateNodeCIDRs`

gmarek commented:
@mqliang - I added a comment in a line where you need to add a guard.

mqliang commented on hack/test-cmd.sh:
```diff
@@ -209,7 +209,8 @@ kube::log::status "Starting controller-manager"
   --kube-api-content-type="${KUBE_TEST_API_TYPE-}" \
-  --master="127.0.0.1:${API_PORT}" 1>&2 &
+  --master="127.0.0.1:${API_PORT}" \
+  --service-cluster-ip-range="10.0.0.0/24" 1>&2 &
```
@gmarek Then, should we remove this line ?

gmarek replied:
Yup.

mqliang committed:
make podcidr mask size configurable

mqliang committed:
update auto generation files

mqliang commented:
@gmarek commits squashed and all test passed. Great thanks for your patient reviewing and guide!

wojtek-t commented:
heh.. @mqliang - can you please rebase?

mqliang commented:
@wojtek-t Rebased.Could you please apply LGTM so we could get this merged ASAP.

wojtek-t commented:
@mqliang - I'm not that familiar with this code, so I will let @gmarek to take a final look

gmarek commented:
I'll do it in half an hour

gmarek commented:
LGTM.

mikedanese commented:
This may have broken GKE CI.

roberthbailey commented:
@andyzheng0831 confirmed that it broke gke ci

wojtek-t commented:
@roberthbailey @mikedanese - I think it was kind of expected that this PR may break things, because we finally implemented feature that we thought we have. So those may be configuration issues with our suites.

andyzheng0831 commented:
It also broke k8s ci running on GCI. I am checking what is wrong. So far it looks like controller-manager process is in crash loop

roberthbailey commented:
That's consistent with what Jenkins shows: `All cluster resources were brought up, but the cluster API is reporting that component "controller-manager" is unhealthy.`

andyzheng0831 commented:
The breakage is fixed by #25984

wojtek-t commented:
@mikedanese - it probably also broke kubemark (at least kubemark-5 started failing around that time, all others didn't finish since the yet, so probably they will also start failing)
I think the problem might b very similar, but I cannot take a look now, can you please take a look?

lavalamp commented:
This PR is no longer automatically revertable...

erictune commented:
@mqliang  Does this PR require action by the user when upgrading from 1.2.x to 1.3.0?  (Think about non-developer users.) If so, please edit your first comment to have  a release-note block, like in https://github.com/kubernetes/kubernetes/pull/28132.  If it is just an optional feature, please change the label to just release-note.  If it is not a complete feature by itself, then apply "release-note-none" label instead.

mqliang commented:
@erictune This is a transparent change for users, it does not require any actions by the user when upgrading. And unfortunately, it seems that I have no permission to change the label.
